File: dashboard/frontend/js/state.js
/**
 * Edge AI Network Anomaly Detection - Centralized Reactive State Store & Event Bus
 * ==============================================================================
 */

class EventBus {
  constructor() {
    this.events = {};
  }

  on(event, listener) {
    if (!this.events[event]) {
      this.events[event] = [];
    }
    this.events[event].push(listener);
    return () => this.off(event, listener);
  }

  off(event, listener) {
    if (!this.events[event]) return;
    this.events[event] = this.events[event].filter((l) => l !== listener);
  }

  emit(event, data) {
    if (!this.events[event]) return;
    this.events[event].forEach((listener) => {
      try {
        listener(data);
      } catch (err) {
        console.error(`Error in event listener for '${event}':`, err);
      }
    });
  }
}

export const eventBus = new EventBus();

export const state = {
  wsConnected: false,
  totalPackets: 0,
  totalThreats: 0,
  currentAnomalyScore: 0.0,
  currentThreatLevel: "NORMAL",
  anomalyThreshold: 0.55,
  connectedNodes: [],
  history: [],
  alerts: [],
  packetFilter: "ALL", // "ALL" | "THREATS"
  audioEnabled: true,
  packetCounter: 1,

  setState(partialState) {
    Object.assign(this, partialState);
    eventBus.emit("stateChanged", this);
  },

  setInitialState(payload) {
    this.totalPackets = payload.total_packets || 0;
    this.totalThreats = payload.total_threats || 0;
    this.anomalyThreshold = payload.anomaly_threshold || 0.55;
    this.connectedNodes = payload.connected_nodes || [];
    this.history = payload.history || [];
    this.alerts = payload.alerts || [];
    this.wsConnected = true;
    eventBus.emit("initialStateLoaded", this);
    eventBus.emit("stateChanged", this);
  },

  addTelemetry(dataPoint, totalPackets, totalThreats) {
    this.totalPackets = totalPackets !== undefined ? totalPackets : this.totalPackets + (dataPoint.packet_rate || 0);
    this.totalThreats = totalThreats !== undefined ? totalThreats : this.totalThreats;
    this.currentAnomalyScore = dataPoint.anomaly_score || 0.0;
    this.currentThreatLevel = dataPoint.severity || "NORMAL";

    this.history.push(dataPoint);
    if (this.history.length > 100) {
      this.history.shift();
    }

    eventBus.emit("telemetryReceived", dataPoint);
    eventBus.emit("stateChanged", this);
  },

  addAlert(alertEntry, totalThreats) {
    this.totalThreats = totalThreats !== undefined ? totalThreats : this.totalThreats + 1;
    this.alerts.push(alertEntry);
    if (this.alerts.length > 50) {
      this.alerts.shift();
    }
    eventBus.emit("alertTriggered", alertEntry);
    eventBus.emit("stateChanged", this);
  },

  updateNode(node) {
    const idx = this.connectedNodes.findIndex((n) => n.device_id === node.device_id);
    if (idx >= 0) {
      this.connectedNodes[idx] = node;
    } else {
      this.connectedNodes.push(node);
    }
    eventBus.emit("nodeUpdated", node);
    eventBus.emit("stateChanged", this);
  }
};

File: dashboard/backend/state.py
"""
Dashboard Backend In-Memory System State
========================================
Quản lý trạng thái luồng telemetry, cảnh báo, danh sách node kết nối trong bộ nhớ.
"""

from typing import List, Dict, Any


class SystemState:
    """Quản lý trạng thái runtime in-memory của hệ thống giám sát."""

    def __init__(self):
        self.connected_nodes: Dict[str, dict] = {}
        self.telemetry_history: List[dict] = []
        self.alerts_history: List[dict] = []
        self.total_packets_inspected: int = 0
        self.total_threats_detected: int = 0
        self.current_anomaly_score: float = 0.0
        self.current_threat_level: str = "NORMAL"
        self.anomaly_threshold: float = 0.55
        self.broker_connected: bool = False

    def add_telemetry(self, data_point: dict):
        self.telemetry_history.append(data_point)
        if len(self.telemetry_history) > 200:
            self.telemetry_history.pop(0)

    def add_alert(self, alert_entry: dict):
        self.total_threats_detected += 1
        self.alerts_history.append(alert_entry)
        if len(self.alerts_history) > 100:
            self.alerts_history.pop(0)

    def get_initial_payload(self) -> dict:
        return {
            "type": "INITIAL_STATE",
            "connected_nodes": list(self.connected_nodes.values()),
            "history": self.telemetry_history[-30:],
            "alerts": self.alerts_history[-20:],
            "total_packets": self.total_packets_inspected,
            "total_threats": self.total_threats_detected,
            "anomaly_threshold": self.anomaly_threshold,
            "broker_connected": self.broker_connected
        }


# Singleton system state instance
state = SystemState()
